Don't assume NV_INF / NV_NAN

The `pkgsrc` repository (a large tree of build instructions for lots of software, used as a package source for NetBSD and many other unix-like distributions) recently updated Perl5 from 5.38 to 5.40, which now no longer builts for VAX targets:
```
gcc -c -DPERL_CORE -O2 -pthread -I/usr/include -fwrapv -fno-strict-aliasing -pipe -std=c99 -O2 -pthread -I/usr/include -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-decl
aration-after-statement -Wc++-compat -Wwrite-strings -DPIC -fPIC builtin.c
In file included from perl.h:6225,
                 from builtin.c:16:
builtin.c: In function 'XS_builtin_inf':  
builtin.c:100:17: error: 'NV_INF' undeclared (first use in this function); did you mean 'NV_MIN'?
  100 |     XSRETURN_NV(NV_INF);
      |                 ^~~~~~
embed.h:649:71: note: in definition of macro 'sv_2mortal'
  649 | # define sv_2mortal(a)                          Perl_sv_2mortal(aTHX_ a)
embed.h:649:71: note: in definition of macro 'sv_2mortal'
  649 | # define sv_2mortal(a)                          Perl_sv_2mortal(aTHX_ a)
      |                                                                       ^
XSUB.h:317:43: note: in expansion of macro 'newSVnv'
  317 | #define XST_mNV(i,v)  (ST(i) = sv_2mortal(newSVnv(v))  )
      |                                           ^~~~~~~
XSUB.h:334:40: note: in expansion of macro 'XST_mNV'
  334 | #define XSRETURN_NV(v)    STMT_START { XST_mNV(0,v);    XSRETURN(1); } STMT_END
      |                                        ^~~~~~~
builtin.c:100:5: note: in expansion of macro 'XSRETURN_NV'
  100 |     XSRETURN_NV(NV_INF);
      |     ^~~~~~~~~~~  
builtin.c:100:17: note: each undeclared identifier is reported only once for each function it appears in
  100 |     XSRETURN_NV(NV_INF);
      |                 ^~~~~~
embed.h:649:71: note: in definition of macro 'sv_2mortal'
  649 | # define sv_2mortal(a)                          Perl_sv_2mortal(aTHX_ a)
      |                                                                       ^
XSUB.h:317:43: note: in expansion of macro 'newSVnv'
  317 | #define XST_mNV(i,v)  (ST(i) = sv_2mortal(newSVnv(v))  )
      |                                           ^~~~~~~
XSUB.h:334:40: note: in expansion of macro 'XST_mNV'
  334 | #define XSRETURN_NV(v)    STMT_START { XST_mNV(0,v);    XSRETURN(1); } STMT_END
      |                                        ^~~~~~~
builtin.c:100:5: note: in expansion of macro 'XSRETURN_NV'
  100 |     XSRETURN_NV(NV_INF);
      |     ^~~~~~~~~~~
builtin.c: In function 'XS_builtin_nan':
builtin.c:110:17: error: 'NV_NAN' undeclared (first use in this function); did you mean 'FP_NAN'?
  110 |     XSRETURN_NV(NV_NAN);
      |                 ^~~~~~
embed.h:649:71: note: in definition of macro 'sv_2mortal'
  649 | # define sv_2mortal(a)                          Perl_sv_2mortal(aTHX_ a)
      |                                                                       ^ 
XSUB.h:317:43: note: in expansion of macro 'newSVnv'
  317 | #define XST_mNV(i,v)  (ST(i) = sv_2mortal(newSVnv(v))  )
      |                                           ^~~~~~~
XSUB.h:334:40: note: in expansion of macro 'XST_mNV'
  334 | #define XSRETURN_NV(v)    STMT_START { XST_mNV(0,v);    XSRETURN(1); } STMT_END
      |                                        ^~~~~~~ 
builtin.c:110:5: note: in expansion of macro 'XSRETURN_NV' 
  110 |     XSRETURN_NV(NV_NAN);
      |     ^~~~~~~~~~~ 
builtin.c: In function 'ck_builtin_const':
builtin.c:138:54: error: 'NV_INF' undeclared (first use in this function); did you mean 'NV_MIN'?
  138 |         case BUILTIN_CONST_INF:   constval = newSVnv(NV_INF); break;
      |                                                      ^~~~~~
embed.h:437:68: note: in definition of macro 'newSVnv'
  437 | # define newSVnv(a)                             Perl_newSVnv(aTHX_ a)
      |                                                                    ^
builtin.c:139:54: error: 'NV_NAN' undeclared (first use in this function); did you mean 'FP_NAN'?
      |                                                                    ^
builtin.c:139:54: error: 'NV_NAN' undeclared (first use in this function); did you mean 'FP_NAN'?
  139 |         case BUILTIN_CONST_NAN:   constval = newSVnv(NV_NAN); break;  
      |                                                      ^~~~~~
embed.h:437:68: note: in definition of macro 'newSVnv'
  437 | # define newSVnv(a)                             Perl_newSVnv(aTHX_ a)
      |                                                                    ^
*** Error code 1
```

This was introduced with 5fdf6e90221900a85cd3091bd2755e41fc230d4d, where INF and NAN support was expected to always exist. However, this is an IEEE fp feature and the VAX CPU predates that spec. Perl used to reasonable support VAX systems (cf. `d_double_style_vax`, `d_double_has_inf` and `d_double_has_nan`.) So I guess that INF/NAN support wasn't tested on a non-IEEE fp system. Reverting this patch makes it build again. Simply guarding that functionality with any of the above config options should just do the trick.

Thanks a lot!

--- a/builtin.c
+++ b/builtin.c
@@ -96,8 +96,12 @@ XS(XS_builtin_inf)
     dXSARGS;
     if(items)
         croak_xs_usage(cv, "");
+#ifdef DOUBLE_HAS_INF
     EXTEND(SP, 1);
     XSRETURN_NV(NV_INF);
+#else
+    Perl_croak_nocontext("builtin::inf not implemented");
+#endif
 }
 
 XS(XS_builtin_nan);
@@ -106,8 +110,12 @@ XS(XS_builtin_nan)
     dXSARGS;
     if(items)
         croak_xs_usage(cv, "");
+#ifdef DOUBLE_HAS_NAN
     EXTEND(SP, 1);
     XSRETURN_NV(NV_NAN);
+#else
+    Perl_croak_nocontext("builtin::nan not implemented");
+#endif
 }
 
 enum {
@@ -135,8 +143,16 @@ static OP *ck_builtin_const(pTHX_ OP *entersubop, GV *namegv, SV *ckobj)
     switch(builtin->ckval) {
         case BUILTIN_CONST_FALSE: constval = &PL_sv_no; break;
         case BUILTIN_CONST_TRUE:  constval = &PL_sv_yes; break;
+#ifdef DOUBLE_HAS_INF
         case BUILTIN_CONST_INF:   constval = newSVnv(NV_INF); break;
+#else
+        case BUILTIN_CONST_INF:   return entersubop;
+#endif
+#ifdef DOUBLE_HAS_NAN
         case BUILTIN_CONST_NAN:   constval = newSVnv(NV_NAN); break;
+#else
+        case BUILTIN_CONST_NAN:   return entersubop;
+#endif
         default:
             DIE(aTHX_ "panic: unrecognised builtin_const value %" IVdf,
                       builtin->ckval);

--- a/lib/builtin.pm
+++ b/lib/builtin.pm
@@ -1,4 +1,4 @@
-package builtin 0.017;
+package builtin 0.018;
 
 use v5.40;
 
@@ -167,7 +167,8 @@ Available starting with Perl 5.36.
 
 This function is currently B<experimental>.
 
-Returns the floating-point infinity value.
+Returns the floating-point infinity value. If the underlying numeric C type
+does not support such a value, it throws a runtime error instead.
 
 Available starting with Perl 5.40.
 
@@ -177,7 +178,8 @@ Available starting with Perl 5.40.
 
 This function is currently B<experimental>.
 
-Returns the floating-point "Not-a-Number" value.
+Returns the floating-point "Not-a-Number" value. If the underlying numeric C
+type does not support such a value, it throws a runtime error instead.
 
 Available starting with Perl 5.40.
 

--- a/lib/builtin.t
+++ b/lib/builtin.t
@@ -8,6 +8,7 @@ BEGIN {
 
 use v5.36;
 no warnings 'experimental::builtin';
+use Config;
 
 package FetchStoreCounter {
     sub TIESCALAR($class, @args) { bless \@args, $class }
@@ -55,19 +56,31 @@ package FetchStoreCounter {
 {
     use builtin qw( inf nan );
 
-    ok(inf, 'inf is true');
-    ok(inf > 1E10, 'inf is bigger than 1E10');
-    ok(inf == inf, 'inf is equal to inf');
-    ok(inf == inf + 1, 'inf is equal to inf + 1');
+    if ($Config{d_double_has_inf}) {
+        ok(inf, 'inf is true');
+        ok(inf > 1E10, 'inf is bigger than 1E10');
+        ok(inf == inf, 'inf is equal to inf');
+        ok(inf == inf + 1, 'inf is equal to inf + 1');
 
-    # Invoke the real XSUB
-    my $inf = ( \&builtin::inf )->();
-    ok($inf == $inf + 1, 'inf returned by real xsub');
+        # Invoke the real XSUB
+        my $inf = ( \&builtin::inf )->();
+        ok($inf == $inf + 1, 'inf returned by real xsub');
+    } else {
+        is(eval { inf }, undef, 'inf throws');
+        my $e = $@;
+        like($e, qr/^builtin::inf not implemented at/, 'inf fails with correct error');
+    }
 
-    ok(nan != nan, 'NaN is not equal to NaN');
+    if ($Config{d_double_has_nan}) {
+        ok(nan != nan, 'NaN is not equal to NaN');
 
-    my $nan = ( \&builtin::nan )->();
-    ok($nan != $nan, 'NaN returned by real xsub');
+        my $nan = ( \&builtin::nan )->();
+        ok($nan != $nan, 'NaN returned by real xsub');
+    } else {
+        is(eval { nan }, undef, 'nan throws');
+        my $e = $@;
+        like($e, qr/^builtin::nan not implemented at/, 'nan fails with correct error');
+    }
 }
 
 # weakrefs

--- a/pod/perldelta.pod
+++ b/pod/perldelta.pod
@@ -129,6 +129,14 @@ L<XXX> has been upgraded from version A.xx to B.yy.
 
 XXX If there was something important to note about this change, include that here.
 
+=item *
+
+L<builtin> has been upgraded from version 0.017 to 0.018.
+
+On platforms that don't support Inf/NaN values in floating-point numbers (such
+as VAX), C<builtin::inf> and C<builtin::nan> now throw a runtime error (rather
+than breaking the perl build). [GH #22882]
+
 =back
 
 =head2 Removed Modules and Pragmata

--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -675,6 +675,16 @@ is currently being compiled. Since this method is used to introduce new
 lexical subroutines into the scope currently being compiled, this is not
 going to have any effect.
 
+=item builtin::inf not implemented
+
+(F) Your machine doesn't support infinity as a numeric value (probably because
+it's a VAX).
+
+=item builtin::nan not implemented
+
+(F) Your machine doesn't support NaN ("not a number") as a numeric value
+(probably because it's a VAX).
+
 =item Builtin version bundle "%s" is not supported by Perl
 
 (F) You attempted to C<use builtin :ver> for a version number that is either

